cmds-42/usr.etc/nmserver/man/nmtest.1
.TH NMTEST 1 2/22/88
.CM 4
.SH NAME
nmtest \- Generic test program for the Mach network server.
.SH SYNOPSIS
.B nmtest
.SH DESCRIPTION
.I nmtest
can be used to test all kinds of message and port transactions with a
.I network server. 
It contains two threads, one to display each message received, and one
to send various messages under control of the user. 

.I nmtest
is interactive; all commands can be listed by entering 
.I ?
or
.I h
at the prompt.
Commands can be in upper or lower case; several commands
can be typed on the same line, provided that spaces are used to 
disambiguate the end of string arguments.
.SH HISTORY
.TP 
22-Feb-88  Daniel Julin (dpj) at Carnegie-Mellon University
Created.


cmds-42/ChangeLog
usr.ucb/from/from.c				mmeyer 8/22/89

Change from to do the from on geteuid(), instead of getlogin() which
looks at who is in /etc/utmp.  

usr.ucb/telnet/telnet.c				mmeyer 8/22/89

Fix for bug #2889.

bin/sed/sed1.c					mmeyer 8/21/89

Fix for bug #2831.

usr.bin/calendar/calendar.{c,sh}		mmeyer 8/21/89

Fix for bug #2794

usr.ucb/Mail					mmeyer 6/27/89

Upgraded to Tahoe sources and fixed a few bugs.

etc/mount/mount.c				mmeyer 6/26/89

Change reversal of mounting of nfs partitions.

etc/init/init.c					mmeyer 6/23/89

Changes for onerror/onoption.

etc/rlogind/rlogind.c				mmeyer 6/14/89

Took out malloc from signal handler.

etc/nu/*					mmeyer 6/13/89

Lee's new nu.

etc/ifconfig/ifconfig.c				mmeyer 6/13/89
etc/bootpd/*

Changes for sexy netinit

bin/awk/awk.g.y					mmeyer 6/1/89

Fixed core dump for Trey.  

etc/quot/quot.c					mmeyer 5/31/89

Changed acct() to print out the correct number of 1024 byte blocks.

bin/sh/homedir.c				mmeyer 5/26/89

Changed sh to use getpwnam() instead of searching through /etc/passwd

usr.ucb/indent/lexi.c				wrp    5/22/89
usr.ucb/indent/indent.c		
usr.ucb/indent/indent_globs.h

Bill Parkhurst's changes for Objective-C 4.0 support, along with
a bugfix that Brian and Morris made.


bin/diff/diffh_vers.c				mike   5/22/89
usr.bin/diff3/Makefile				

Bugfix from Mike DeMoney.  Fixes problem that occurs during integration.

usr.bin/tip/acu.c				mmeyer 5/20/89
usr.bin/tip/cu.c
usr.bin/tip/tip.c
usr.bin/tip/cmds.c

Changed references to connect(), abort() and send() to compile
with shlibs.

bin/csh/sh.c					mmeyer 5/19/89

Integrated the atexit() mod.

usr.ucb/rlogin/rlogin.c				mmeyer 5/19/89

Verified that Mike DeMoney's change to make chars with 0x7f was in.

bin/make/rules.c				mmeyer 5/19/89
bin/make/doname.c	

Integrated Mike's fixes.

usr.etc/mach_swapon/swaptab			mmeyer 5/19/89

Changed to /private/vm/swaptab instead of /etc/vm/swaptab

usr.bin/calendar/Makefile			mmeyer 5/18/89

Changed Makefile so /usr/lib/calendar would install correctly.

etc/route/route.c				mmeyer 5/16/89
etc/rwhod/rwhod.c
usr.etc/nfsstat/nfsstat.c
usr.etc/rpc.rstatd/rpc.rstatd.c
usr.ucb/sysline/sysline.c (not that this will ever run on our system)
usr.ucb/sendmail/src/conf.c
usr.bin/uucp/acucntrl.c

Changed all to nlist "/mach" instead of "/vmunix"

etc/syslog/syslog.c				mmeyer 5/15/89

Changed syslog to print out "mach:" instead of "vmunix:"

bin/date/date.c					mmeyer 5/14/89

Initialized tz.{tz_minuteswest,tz_dsttime} to zero.  This bug
cropped up when linking with shared libraries.  Kevin Enderby
has a bug report.

mk/cmds/usr/ucb/w/Makefile			mmeyer 5/14/89

Linked uptime to w in the INSTALL phase

usr.bin/uucp/fio.c				mmeyer 5/9/89
usr.bin/uucp/gio.c
usr.bin/uucp/logent.c
usr.bin/uucp/tio.c

Ifdef'd syslog() so uucp would link with the libsys_s shlib.
	

usr.bin/uucp/Makefile

Took out duplicate index.c from uucp.

usr.bin/troff/Makefile				mmeyer 5/9/89

Took out malloc.c and calloc.c from CFILES.  troff had its own 
version of calloc,realloc,malloc and free.

usr.bin/nroff/Makefile				mmeyer 5/9/89

Took out malloc.c and calloc.c from CFILES.  nroff had its own 
version of calloc,realloc,malloc and free.

usr.ucb/rusers/Makefile				mmeyer 5/9/89

Took out malloc.c and calloc.c from CFILES.  rusers had its own 
version of calloc,realloc,malloc and free.

usr.ucb/msgs/msgs.c	  			mmeyer 5/9/89

Ifdef'd send() so /usr/ucb/msgs would link with the libsys_s shlib.

usr.ucb/more/more.c	  			mmeyer 5/9/89

Ifdef'd strlen() so /usr/ucb/more would link with the libsys_s shlib.

usr.ucb/ex/exrecover.c  			mmeyer 5/9/89

Ifdef'd fprintf() so /usr/ucb/exrecover would link with the libsys_s shlib.

usr.ucb/Mail/aux.c  				mmeyer 5/9/89

Ifdef'd index() so /usr/ucb/Mail would link with the libsys_s shlib.

usr.ucb/Mail/send.c				mmeyer 5/9/89
usr.ucb/Mail/cmd1.c
usr.ucb/Mail/cmd2.c
usr.ucb/Mail/collect.c
usr.ucb/Mail/edit.c
usr.ucb/Mail/fio.c
usr.ucb/Mail/quit.c

Ifdef'd send() so /usr/ucb/Mail would link with the libsys_s shlib.

usr.ucb/Mail/head.c
usr.ucb/Mail/list.c
usr.ucb/Mail/aux.c

Ifdef'd raise() so /usr/ucb/Mail would link with the libsys_s shlib.

etc/sa/sa.c					mmeyer 5/9/89

Ifdef'd bcmp() so /etc/sa would link with the libsys_s shlib.

etc/routed/defs.h				mmeyer 5/9/89
etc/routed/startup.c
etc/routed/inet.c 

Ifdef'd inet_makeaddr() so /etc/routed would link with the libsys_s shlib.

bin/su/su.c					mmeyer 5/9/89

Unifdef'd getenv() so su would link with the libsys_s shlib.

bin/awk/Makefile				mmeyer 5/9/89

Took out malloc.c and calloc.c from CFILES.  awk had its own 
version of calloc,realloc,malloc and free.

All CMDS Makefiles				mmeyer 5/8/89

Changed all Makefiles to have -O -g for CFLAGS and -lsys_s for
LIBS.  Changes above are for commands that didn't link.

cmds/bin/cp/cp.c				mmeyer 5/8/89
cmds/bin/cat/cat.c

Integrated Avie's mapped file changes.  He claims a 10% performace
win with mapped files.

cmds/bin/make/main.c				mmeyer 5/1/89

Changed main() to have make to a setrlimit(RLIMIT_STACK) to trap
infinitely recursive macros instead of core dumping.

cmds/bin/sh/homedir.c				mmeyer  4/30/89

Changed sh to use getpwent() rather than grep'ing through /etc/passwd.

cmds/bin/tar/tar.c				mmeyer  4/30/89

Changed the "v" flag printf()'s to fprintf(stderr)'s

cmds/etc/icheck/icheck.c			mmeyer  4/30/89

Modified icheck to know about fast symlinks.

cmds/usr.ucb/rlogin/rlogin.c			mmeyer

Added the oob() signal handler for the SIGURG signal.  Data
was getting lost between the socket and the rlogin program.

cmds/bin/csh/sh.c				mmeyer

Added lines 1022 to 1029 in dosource() which check for no operand 
for source flag.

Changed lines 1036 to 1042 in dosource() to just print out an error
using perror() when an open of a source'd file fails instead of not
letting the login session continue.

cmds/bin/passwd/passwd.c			mmeyer

Took out a check against /bin/sh so a user can switch back
and forth from csh to sh and back again.


cmds-42/usr.etc/nmserver/utils/nmtest.c
/* 
 * Mach Operating System
 * Copyright (c) 1987 Carnegie-Mellon University
 * All rights reserved.  The CMU software License Agreement specifies
 * the terms and conditions for use and redistribution.
 */
/*
 * nmtest.c
 *
 * $Source: /os/osdev/LIBS/libs-7/etc/nmserver/utils/RCS/nmtest.c,v $
 *
 */
#ifndef	lint
static char     rcsid[] = "$Header: nmtest.c,v 1.1 88/09/30 15:47:07 osdev Exp $";
#endif not lint
/*
 * Program to test various operations of the network server.
 */

/*
 * HISTORY: 
 * 19-Sep-87  Daniel Julin (dpj) at Carnegie-Mellon University
 *	Created.
 *
 */


#include	"debug.h"
#include	"trace.h"
#include	"nm_defs.h"

#include	<stdio.h>
#include	<mach.h>
#include	<cthreads.h>
#include	<servers/netname.h>
#include	<sys/message.h>
#include	<msg_type.h>
#include	<sys/notify.h>

#define readln(p) while (((getc(p))!='\n') && (!feof(p)))
#define NETNAME_NAME	"NEW_NETWORK_NAME_SERVER"

#define	MAXPORTS	40

#define	RIGHT_SEND	0x1
#define	RIGHT_RECEIVE	0x2
#define	RIGHT_OWNER	0x4


/*
 * Global variables
 */
static port_t		network_server_port;		/* port to new network server */
static struct ports {
	port_t	port;
	char	host[100];
	char	name[100];
	long	rights;
} ports[MAXPORTS];
int	last_port;
struct	my_msg {
	msg_header_t	header;
	msg_type_long_t	desc;
	long		data;
} send_buff;
long	current_seqnum;
char	rcv_msg[100];

#define	CHECK_PORT(index,stmt) {						\
	if ((index < 0) || (index > last_port) || (ports[index].rights == 0)) {	\
		fprintf(stdout,"Invalid port number: %d\n", index);		\
		stmt;								\
	}									\
}

/*
 * Tracing values.
 */
int	trace_recursion_level = 0;
#undef	tracing_on
int	tracing_on = 0;

/*
 * input
 *
 * Obtain user input, allowing multiple inputs on the same line.
 *
 * Parameters:
 *
 * prompt: string to use as a prompt.
 * format: specification for data required.
 * OUT var: address where to put the data read.
 *
 * Results:
 *
 * Side effects:
 *
 * May print a prompt on stdout, and start a read on stdin.
 *
 * Design:
 *
 * Note:
 *
 */
void input(prompt, format, var)
char *prompt;
char *format;
char *var;
BEGIN("input")
	char c;
	int fcnt;

	while ((stdin->_cnt > 0) &&
	    ((((c = *(stdin)->_ptr&0377) == ' ') || (c == '\n'))))
		getc(stdin);
	fcnt = stdin->_cnt;

	if (fcnt == 0) {
		fprintf(stdout, "%s", prompt);
		fflush(stdout);
	}
	fscanf(stdin, format, var);

	RET;
END


/*
 * nmconnnect
 *
 * Connect to the network name server.
 */
void nmconnect()
BEGIN("nmconnect")
	kern_return_t	kr;

	kr = netname_look_up(name_server_port, "", NETNAME_NAME, &network_server_port);
	if (kr == NETNAME_SUCCESS) {
		fprintf(stdout,
			"Look up of new network server succeeds, port = %x.\n",
			network_server_port);
	} else {
		fprintf(stdout,
	"Look up of new network server fails, using old network server, port = %x.\n",
			name_server_port);
		network_server_port = name_server_port;
	}
	RET;
END


/*
 * printmenu
 *
 * Parameters: none
 *
 * Results:
 *
 * Side effects:
 *
 * Print a command menu on stdout.
 *
 * Design:
 *
 * Note:
 *
 */
void printmenu()
BEGIN("printmenu")
	fprintf(stdout, "\n");
	fprintf(stdout, "This programs allows to perform tests on a network server.\n");
	fprintf(stdout, "\n");
	fprintf(stdout, "Available commands:\n");
	fprintf(stdout, "\n");
	fprintf(stdout, "    H,?.......Print this message\n");
	fprintf(stdout, "    C.........(Re)connect to network server\n");
	fprintf(stdout, "\n");
	fprintf(stdout, " Ports:\n");
	fprintf(stdout, "       PA.........Allocate\n");
	fprintf(stdout, "       PC,PI......Check in <port> <name>\n");
	fprintf(stdout, "       PO.........Check out <port>\n");
	fprintf(stdout, "       PL.........Look up <host> <name>\n");
	fprintf(stdout, "       PD.........Deallocate <port>\n");
	fprintf(stdout, "       PS.........Print status of all known ports\n");
	fprintf(stdout, "\n");
	fprintf(stdout, " Messages:\n");
	fprintf(stdout, "       MI.........Send IPC <remote-port> <local-port>\n");
	fprintf(stdout, "       MR.........Send RPC <remote-port> <local-port>\n");
	fprintf(stdout, "\n");
	fprintf(stdout, " Rights:\n");
	fprintf(stdout, "       RS.........Transfer send rights <port> <destination>\n");
#ifdef MSG_TYPE_PORT_RECEIVE
	fprintf(stdout, "       RR.........Transfer receive rights <port> <destination>\n");
#endif	MSG_TYPE_PORT_RECEIVE
#ifdef	MSG_TYPE_PORT_OWNERSHIP
	fprintf(stdout, "       RO.........Transfer ownership rights <port> <destination>\n");
#endif	MSG_TYPE_PORT_OWNERSHIP
	fprintf(stdout, "       RA.........Transfer all rights <port> <destination>\n");
	fprintf(stdout, "\n");
	fprintf(stdout, "    Q.........Quit\n");
	fprintf(stdout, "\n");
	RET;
END


/*
 * printport
 *
 * Print all information on a port.
 */
void printport(str,index)
	char	*str;
	int	index;
BEGIN("printport")
	CHECK_PORT(index,return);
	fprintf(stdout,"%20s %2d: 0x%8x ",str,index,ports[index].port);
	if (ports[index].rights & RIGHT_SEND)
		fprintf(stdout,"SEND ");
	else
		fprintf(stdout,"     ");
	if (ports[index].rights & RIGHT_RECEIVE)
		fprintf(stdout,"RECEIVE ");
	else
		fprintf(stdout,"        ");
	if (ports[index].rights & RIGHT_OWNER)
		fprintf(stdout,"OWNER ");
	else
		fprintf(stdout,"      ");
	fprintf(stdout," \"%s\" \"%s\"\n",ports[index].host,ports[index].name);
	RET;
END


/*
 * findport
 *
 * Make sure there is a record for a given port.
 */
int findport(port)
	port_t	port;
BEGIN("findport")
	int	port_index;
	int	i;

	port_index = -1;
	for (i = 0; i < MAXPORTS; i++) {
		if (ports[i].port == port) {
			port_index = i;
			break;
		}
	}
	if (port_index == -1) {
		if (last_port >= (MAXPORTS - 1)) {
			port_index = 0;
			fprintf(stdout,"** Sorry, you have exceeded the maximum **\n");
			fprintf(stdout,"** number of ports for this program.    **\n");
			fflush(stdout);
		} else {
			last_port++;
			port_index = last_port;
			ports[port_index].port = port;
			ports[port_index].host[0] = '\0';
			ports[port_index].name[0] = '\0';
			ports[port_index].rights = 0;
		}
	}

	RETURN(port_index);
END


/*
 * genericports
 *
 * Execute any port operation.
 */
void genericports()
BEGIN("genericports")
	char 		subcmd[2];
	kern_return_t	kr;
	int		port_index;
	char		host[100];
	char		name[100];
	port_t		new_port;

	input("Enter Port sub-command: ", "%1s", subcmd);
	switch(subcmd[0]) {
		case 'A':
		case 'a':
			kr = port_allocate(task_self(),&new_port);
			if (kr == KERN_SUCCESS) {
				port_index = findport(new_port);
				strcpy(ports[port_index].host,"local");
				ports[port_index].rights = 
					RIGHT_SEND | RIGHT_RECEIVE | RIGHT_OWNER;
				(void)port_enable(task_self(),ports[port_index].port);
				printport("Allocated port",port_index);
			} else {
				fprintf(stdout,"port_allocate returns %d\n",kr);
			}
			break;
		case 'C':
		case 'c':
		case 'I':
		case 'i':
			input("Port: ", "%d", (char *)&port_index);
			CHECK_PORT(port_index,break);
			input("Name: ", "%99s", name);
			kr = netname_check_in(network_server_port, name,
						PORT_NULL, ports[port_index].port);
			if (kr == NETNAME_SUCCESS) {
				fprintf(stdout,"Check in succeeds.\n");
				strcpy(ports[port_index].host,"local");
				strcpy(ports[port_index].name,name);
			} else {
				fprintf(stdout,"netname_check_in returns %d\n",kr);
			}
			break;
		case 'O':
		case 'o':
			input("Port: ", "%d", (char *)&port_index);
			CHECK_PORT(port_index,break);
			kr = netname_check_out(network_server_port, 
						ports[port_index].name,PORT_NULL);
			if (kr == NETNAME_SUCCESS) {
				fprintf(stdout,"Check out succeeds.\n");
				ports[port_index].name[0] = '\0';
			} else {
				fprintf(stdout,"netname_check_out returns %d\n",kr);
			}
			break;
		case 'L':
		case 'l':
			input("Host: ", "%99s", host);
			input("Name: ", "%99s", name);
			kr = netname_look_up(network_server_port, host, name, &new_port);
			if (kr == NETNAME_SUCCESS) {
				fprintf(stdout,"Look up succeeds:\n");
				port_index = findport(new_port);
				strcpy(ports[port_index].host,host);
				strcpy(ports[port_index].name,name);
				ports[port_index].rights |= RIGHT_SEND;
				printport("Found port",port_index);
			} else {
				fprintf(stdout,"netname_look_up returns %d\n",kr);
			}
			break;
		case 'D':
		case 'd':
			input("Port: ", "%d", (char *)&port_index);
			CHECK_PORT(port_index,break);
			kr = port_deallocate(task_self(),ports[port_index].port);
			if (kr == KERN_SUCCESS) {
				fprintf(stdout,"Port 0x%x deallocated.\n",
							ports[port_index].port);
				ports[port_index].rights = 0;
			} else {
				fprintf(stdout,"port_deallocate returns %d\n",kr);
			}
			break;
		case 'S':
		case 's':
			for (port_index = 0; port_index < MAXPORTS; port_index++) {
				if (ports[port_index].rights != 0)
					printport("Port",port_index);
			}
			break;
		default:
			fprintf(stdout, "Invalid subcommand. Type ? for help.\n");
			readln(stdin);
			break;
	}
	fflush(stdout);
	RET;
END


/*
 * genericmsg
 *
 * Execute any message operation.
 */
void genericmsg()
BEGIN("genericmsg")
	char		subcmd[2];
	int		msg_type;
	int		remote_port;
	int		local_port;
	kern_return_t	kr;

	input("Enter Message sub-command: ", "%1s", subcmd);
	switch(subcmd[0]) {
		case 'I':
		case 'i':
			msg_type = MSG_TYPE_NORMAL;
			break;
		case 'R':
		case 'r':
			msg_type = MSG_TYPE_RPC;
			break;
		default:
			fprintf(stdout, "Invalid subcommand. Type ? for help.\n");
			readln(stdin);
			RET;
	}

	input("Remote Port: ", "%d", (char *)&remote_port);
	input("Local Port: ", "%d", (char *)&local_port);
	CHECK_PORT(remote_port,return);
	CHECK_PORT(local_port,return);

	send_buff.header.msg_simple = TRUE;
	send_buff.header.msg_size = sizeof(struct my_msg);
	send_buff.header.msg_type = msg_type;
	send_buff.header.msg_local_port = ports[local_port].port;
	send_buff.header.msg_remote_port = ports[remote_port].port;
	send_buff.header.msg_id = current_seqnum;

	send_buff.desc.msg_type_long_name = MSG_TYPE_INTEGER_32;
	send_buff.desc.msg_type_long_size = 32;
	send_buff.desc.msg_type_long_number = 1;
	send_buff.desc.msg_type_header.msg_type_inline = TRUE;
	send_buff.desc.msg_type_header.msg_type_longform = TRUE;
	send_buff.desc.msg_type_header.msg_type_deallocate = FALSE;
	send_buff.data = current_seqnum;

	kr = msg_send((msg_header_t *)&send_buff, MSG_OPTION_NONE, 0);
	if (kr == SEND_SUCCESS) {
		fprintf(stdout,"Message away,seqnum = %d\n",current_seqnum);
	} else {
		fprintf(stdout,"msg_send returned %d\n",kr);
	}
	current_seqnum++;

	fflush(stdout);
	RET;
END


/*
 * genericrights
 *
 * Execute any rights operation.
 */
void genericrights()
BEGIN("genericrights")
	char		subcmd[2];
	int		rights;
	int		rightmask;
	int		destination_port;
	int		object_port;
	kern_return_t	kr;

	input("Enter Rights sub-command: ", "%1s", subcmd);
	switch(subcmd[0]) {
		case 'S':
		case 's':
			rights = MSG_TYPE_PORT;
			rightmask = 0;
			break;
#ifdef	MSG_TYPE_PORT_RECEIVE
		case 'R':
		case 'r':
			rights = MSG_TYPE_PORT_RECEIVE;
			rightmask = RIGHT_RECEIVE;
			break;
#endif	MSG_TYPE_PORT_RECEIVE
#ifdef	MSG_TYPE_PORT_OWNERSHIP
		case 'O':
		case 'o':
			rights = MSG_TYPE_PORT_OWNERSHIP;
			rightmask = RIGHT_OWNER;
			break;
#endif	MSG_TYPE_PORT_OWNERSHIP
		case 'A':
		case 'a':
			rights = MSG_TYPE_PORT_ALL;
			rightmask = RIGHT_RECEIVE;
			break;
		default:
			fprintf(stdout, "Invalid subcommand. Type ? for help.\n");
			readln(stdin);
			RET;
	}

	input("Object Port: ", "%d", (char *)&object_port);
	input("Destination Port: ", "%d", (char *)&destination_port);
	CHECK_PORT(object_port,return);
	CHECK_PORT(destination_port,return);

	send_buff.header.msg_simple = FALSE;
	send_buff.header.msg_size = sizeof(struct my_msg);
	send_buff.header.msg_type = MSG_TYPE_NORMAL;
	send_buff.header.msg_local_port = PORT_NULL;
	send_buff.header.msg_remote_port = ports[destination_port].port;

	send_buff.desc.msg_type_long_name = rights;
	send_buff.desc.msg_type_long_size = 32;
	send_buff.desc.msg_type_long_number = 1;
	send_buff.desc.msg_type_header.msg_type_inline = TRUE;
	send_buff.desc.msg_type_header.msg_type_longform = TRUE;
	send_buff.desc.msg_type_header.msg_type_deallocate = FALSE;
	send_buff.data = (long)ports[object_port].port;

	kr = msg_send((msg_header_t *)&send_buff, MSG_OPTION_NONE, 0);
	if (kr == SEND_SUCCESS) {
		ports[object_port].rights &= ~rightmask;
	} else {
		fprintf(stdout,"msg_send returned %d\n",kr);
	}

	fflush(stdout);
	RET;
END


/*
 * main
 *
 * Main loop.
 *
 * Parameters:
 *
 * Results:
 *
 * Side effects:
 *
 * Design:
 *
 * Note:
 *
 */
main(argc, argv)
int argc;
char **argv;
BEGIN("main")
	int quit;
	char cmd[2];

	(void)initialize();

	quit = FALSE;
	do {
		input("> ", "%1s", cmd);
		switch (cmd[0]) {
		case 'H':
		case 'h':
		case '?':
			printmenu();
			break;
		case 'C':
		case 'c':
			nmconnect();
			break;
		case 'P':
		case 'p':
			genericports();
			break;
		case 'M':
		case 'm':
			genericmsg();
			break;
		case 'R':
		case 'r':
			genericrights();
			break;
		case 'Q':
		case 'q':
			quit = TRUE;
			break;
		default:
			fprintf(stdout, "Invalid command. Type ? for list.\n");
			readln(stdin);
			break;
		}
	} 
	while (!quit);

	fprintf(stdout, "Terminated.\n");
	exit(0);
END


/*
 * receive_main
 *
 * Main loop for the msg receive thread.
 */
void receive_main()
BEGIN("receive_main")
	msg_return_t	msg_ret;
	int		index;
	struct my_msg	*rcv_buff;
	notification_t	*notify_buff;

	for (;;) {
		rcv_buff = (struct my_msg *)&rcv_msg;
		rcv_buff->header.msg_local_port = PORT_ENABLED;
		rcv_buff->header.msg_size = sizeof(struct my_msg);
		rcv_buff->desc.msg_type_long_name = 0;
		rcv_buff->data = 0;

		msg_ret = msg_receive(rcv_buff,MSG_OPTION_NONE,0);
		if (msg_ret != RCV_SUCCESS) {
			fprintf(stdout,"msg_receive returned %d\n",msg_ret);
			fflush(stdout);
			continue;
		}

		if (rcv_buff->header.msg_local_port == task_notify()) {
			fprintf(stdout,"\nReceived a notification:\n");
			notify_buff = (notification_t *)rcv_buff;
			index = findport((port_t)notify_buff->notify_port);
			switch(notify_buff->notify_header.msg_id) {
				case NOTIFY_RECEIVE_RIGHTS:
					ports[index].rights |= 
							RIGHT_RECEIVE | RIGHT_SEND;
					(void)port_enable(task_self(),ports[index].port);
					printport(" Receive rights",index);
					break;
				case NOTIFY_OWNERSHIP_RIGHTS:
					ports[index].rights |= RIGHT_OWNER | RIGHT_SEND;
					printport(" Ownership rights",index);
					break;
				case NOTIFY_PORT_DELETED:
					printport(" Port dead",index);
					ports[index].rights = 0;
					break;
				default:
					fprintf(stdout,"Unknown notification: %d\n",
						notify_buff->notify_header.msg_id);
					break;
			}
		} else {
			fprintf(stdout,"\nReceived a message:\n");
			index = findport(rcv_buff->header.msg_local_port);
			ports[index].rights |= RIGHT_RECEIVE;
			printport("Local port",index);
			index = findport(rcv_buff->header.msg_remote_port);
			ports[index].rights |= RIGHT_SEND;
			printport("Remote port",index);
			switch(rcv_buff->desc.msg_type_long_name) {
				case MSG_TYPE_INTEGER_32:
					if (rcv_buff->header.msg_type & MSG_TYPE_RPC) {
						fprintf(stdout,"   RPC  ");
					} else {
						fprintf(stdout,"        ");
					}
					fprintf(stdout,"  Data = %d\n",
									rcv_buff->data);
					break;
				case MSG_TYPE_PORT:
					index = findport((port_t)rcv_buff->data);
					ports[index].rights |= RIGHT_SEND;
					printport(" Send rights",index);
					break;
#if	MSG_TYPE_PORT_RECEIVE
				case MSG_TYPE_PORT_RECEIVE:
					index = findport((port_t)rcv_buff->data);
					ports[index].rights |= 
							RIGHT_RECEIVE | RIGHT_SEND;
					(void)port_enable(task_self(),ports[index].port);
					printport(" Receive rights",index);
					break;
#endif	MSG_TYPE_PORT_RECEIVE
#if	MSG_TYPE_PORT_OWNERSHIP
				case MSG_TYPE_PORT_OWNERSHIP:
					index = findport((port_t)rcv_buff->data);
					ports[index].rights |= RIGHT_OWNER | RIGHT_SEND;
					printport(" Ownership rights",index);
					break;
#endif	MSG_TYPE_PORT_OWNERSHIP
				case MSG_TYPE_PORT_ALL:
					index = findport((port_t)rcv_buff->data);
					ports[index].rights |= 
						RIGHT_RECEIVE | RIGHT_SEND;
					(void)port_enable(task_self(),ports[index].port);
					printport(" All rights",index);
					break;
				default:
					fprintf(stdout,"Unknown message data: %d\n",
						rcv_buff->desc.msg_type_long_name);
					break;
			}
		}
		fflush(stdout);
	}
	/* NOTREACHED */
END


/*
 * initialize
 *
 * Initialize all the test functions.
 */
int initialize()
BEGIN("initialize")
	int		i;
	cthread_t	receive_thread;

	/*
	 * Connect to name server.
	 */
#ifndef	NeXT
	init_netname(PORT_NULL);
#endif	NeXT
	nmconnect();

	current_seqnum = 1;

	/*
	 * Prepare port list.
	 */
	for (i = 0; i < MAXPORTS; i++)
		ports[i].rights = 0;
	last_port = 0;
	ports[0].port = PORT_NULL;
	ports[0].host[0] = '\0';
	strcpy(ports[0].name,"NULLPORT");
	ports[0].rights = RIGHT_SEND;

	/*
	 * Set up the msg receive thread.
	 */
	cthread_init();
	receive_thread = cthread_fork(receive_main, 0);
	cthread_set_name(receive_thread, "receive_main");
    	cthread_detach(receive_thread);

	(void)port_enable(task_self(),task_notify());

	RETURN(1);
END



